src/Database/LSMTree/Internal.hs

```diff
@@ -1344,6 +1394,57 @@ openSnapshot sesh label tableType override snap resolve = do
           , tableUnionLevel = unionLevel
           }
 
+{-# SPECIALISE wrapFileCorruptedErrorAsSnapshotCorruptedError ::
+       SnapshotName
+    -> IO a
+    -> IO a
+    #-}
+wrapFileCorruptedErrorAsSnapshotCorruptedError ::
+       forall m a.
+       (MonadCatch m)
+    => SnapshotName
+    -> m a
+    -> m a
+wrapFileCorruptedErrorAsSnapshotCorruptedError snapshotName action =
+    action `catches` handlers
+    where
+        handlers :: [Handler m a]
+        handlers =
+            [ Handler $ throwIO . wrapFileCorruptedError
+            , Handler $ throwIO . wrapAbortActionRegistryError
+            , Handler $ throwIO . wrapCommitActionRegistryError
+            ]
+
+        -- TODO: This erases the `ExceptionContext` of the underlying `FileCorruptedError`,
+        --       `AbortActionRegistryError`, or `CommitActionRegistryError`.
+        --       Unfortunately, the API exposed by `io-classes` does not currently expose
+        --       any primitives that could be used to preserve the `ExceptionContext`.
+        wrapSomeException :: SomeException -> SomeException
+        wrapSomeException e =
+            fromMaybe e . getFirst . mconcat . fmap First $
+                [ toException . wrapFileCorruptedError <$> fromException e
+                , toException . wrapAbortActionRegistryError <$> fromException e
+                , toException . wrapCommitActionRegistryError <$> fromException e
+                ]
+
+        wrapFileCorruptedError :: FileCorruptedError -> SnapshotCorruptedError
+        wrapFileCorruptedError = ErrSnapshotCorrupted snapshotName
```

Comment:
So the user would see an `AbortActionRegistryError` or `CommitActionRegistryError` that contains an `ErrSnapshotCorrupted`?

I think the nicer thing would be if we flip this: the user gets `ErrSnapshotCorrupted` containing a `FileCorruptedError`/`AbortActionRegistryError`/`CommitActionRegistryError`

Reply:
The point is only to wrap the file corruption error in a snapshot corruption error, not to restructure anything.

Reply:
I suppose I'm suggesting to restructure it. Would it be something you think makes sense to do? Currently, I wouldn't expect users to have to catch `ActionRegistry` exceptions, but maybe you had a design in mind?

Reply:
I suggest that this change should be considered out-of-scope for this PR, and be the subject of a future PR.

It's entirely non-obvious to me how to decide which of the various exceptions under an ActionRegistry exception should take precedence, e.g., what do you do if there's two `FileCorruptedError`s?

Reply:
Okay, let's do it in a follow-up PR

> It's entirely non-obvious to me how to decide which of the various exceptions under an ActionRegistry exception should take precedence, e.g., what do you do if there's two `FileCorruptedError`s?

Probably in chronological order

Reply:
Makes sense, but might be better to present the user with the whole picture rather than selectively hide stuff and possibly get that wrong.

Reply:
Might be a good change, but let's push that for a future PR.

Reply:
See #646 for follow-up discussion.